libnetwork: share a single datastore with drivers

corhere opened the pull request:
The bbolt library wants exclusive access to the boltdb file and uses file locking to assure that is the case. The controller and each network driver that needs persistent storage instantiates its own unique datastore instance, backed by the same boltdb file. The boltdb kvstore implementation works around multiple access to the same boltdb file by aggressively closing the boltdb file between each transaction. This is very inefficient. Have the controller pass its datastore instance into the drivers and enable the PersistConnection option to disable closing the boltdb between transactions.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**
Watch CI and hope for the best

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
libnetwork: share a single datastore with drivers

The bbolt library wants exclusive access to the boltdb file and uses
file locking to assure that is the case. The controller and each network
driver that needs persistent storage instantiates its own unique
datastore instance, backed by the same boltdb file. The boltdb kvstore
implementation works around multiple access to the same boltdb file by
aggressively closing the boltdb file between each transaction. This is
very inefficient. Have the controller pass its datastore instance into
the drivers and enable the PersistConnection option to disable closing
the boltdb between transactions.

Set data-dir in unit tests which instantiate libnetwork controllers so
they don't hang trying to lock the default boltdb database file.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented:
Looks like some failures are consistent between graphdriver and snapshotter runs may be legit; 


```
=== Failed
=== FAIL: daemon TestIpcPrivateVsReadonly (59.98s)
    oci_linux_test.go:111: assertion failed: error is not nil: timeout

=== FAIL: daemon TestSysctlOverride (59.95s)
    oci_linux_test.go:139: assertion failed: error is not nil: timeout

=== FAIL: daemon TestSysctlOverrideHost (60.00s)
    oci_linux_test.go:189: assertion failed: error is not nil: timeout

=== FAIL: daemon TestDefaultResources (59.99s)
    oci_linux_test.go:226: assertion failed: error is not nil: timeout

=== FAIL: daemon TestDaemonReloadNetworkDiagnosticPort (59.97s)
    reload_test.go:369: timeout

=== FAIL: daemon TestGetBlkioWeightDevices (unknown)

=== FAIL: daemon TestGetBlkioThrottleDevices (unknown)

=== FAIL: daemon TestExecResizeTimeout (unknown)
coverage: 13.3% of statements
```

thaJeztah commented:
No clue why they fail on this PR though 🤔 Let me restart CI

corhere commented:
Nah, those failures are legit. There's nothing (necessarily) wrong with the code, but the tests need to be changed so as not to instantiate libnetwork controllers with the default datastore path

thaJeztah commented:
Yeah, I later noticed that one of the tests had a stack related to libnetwork, so perhaps that affects the other tests.

corhere committed:
libnetwork/datastore: don't parse file path

File paths can contain commas, particularly paths returned from
t.TempDir() in subtests which include commas in their names. There is
only one datastore provider and it only supports a single address, so
the only use of parsing the address is to break tests in mysterious
ways.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented:
Oh! Nice and shiny green now ✅ 🥳 


@akerouanton @robmry ptal 🤗

robmry reviewed (approved):
LGTM

akerouanton commented on libnetwork/datastore/datastore.go:
```diff
@@ -93,6 +92,7 @@ func DefaultScope(dataDir string) ScopeCfg {
 			Config: &store.Config{
 				Bucket:            "libnetwork",
 				ConnectionTimeout: time.Minute,
+				PersistConnection: true,
```
Do you plan to open a follow-up PR to remove this option?

corhere replied:
Eventually, I guess. Do you want to do it?

akerouanton replied:
- https://github.com/moby/moby/pull/47302

akerouanton commented on libnetwork/drivers/bridge/bridge_store.go:
```diff
@@ -25,17 +24,13 @@ const (
 
 func (d *driver) initStore(option map[string]interface{}) error {
 	if data, ok := option[netlabel.LocalKVClient]; ok {
```
It'd be cleaner to change drivers' `Register` function to add a dedicated arg for passing the datastore. Is this something you're planning to do in a follow-up?

corhere replied:
More type-safe, sure, but that'd just be changing the shape of dependency injection. I don't think it's high priority; I'd like to leave it alone for now and wait for a wider refactoring of the driver interface to make such a change.

akerouanton reviewed (approved):
LGTM
